File: src/shared/render/clipping.ts
/**
 * Clipping — what has gone to white or to black, pixel by pixel (audit item
 * 15, `docs/lightroom-gaps.md`): the rule the histogram's clip marks count by,
 * the colours the stage paints it in, and how a readout tells a painted pixel
 * from a photographed one.
 *
 * ONE rule for the three readers — the strip's percentages
 * (`develop/histogram.ts`), the overlay on the picture (`clip-pass.ts`, a
 * transcription of `clipOf`) and the readout under the pointer — so a pixel
 * the strip counts is a pixel the picture marks.
 *
 * Pure and DOM-free.
 */

/** An 8-bit value at or past which a channel has lost its highlight detail. */
export const CLIP_WHITE = 254;
/** An 8-bit value at or under which a channel has lost its shadow detail. */
export const CLIP_BLACK = 1;

export type Clip = 'white' | 'black';

/**
 * Whether an 8-bit pixel is clipped: to WHITE when ANY channel is — a sky
 * whose red alone has gone has lost its colour, and that is what a
 * photographer means by blown — and to BLACK only when EVERY channel is,
 * because a saturated blue with no red in it is a colour, not a crushed
 * shadow. White is asked first: a pixel cannot be both.
 */
export function clipOf(r: number, g: number, b: number): Clip | null {
  if (r >= CLIP_WHITE || g >= CLIP_WHITE || b >= CLIP_WHITE) return 'white';
  if (r <= CLIP_BLACK && g <= CLIP_BLACK && b <= CLIP_BLACK) return 'black';
  return null;
}

/**
 * The colours a clipped pixel is painted in on the stage — Lightroom's own
 * (red for highlights, blue for shadows), because a photographer reads them
 * without a legend.
 *
 * Each has ONE channel at 255, which is what lets `readoutOf` know it: a
 * photographed pixel within a step of either colour has that channel at 254
 * or more, so it is itself clipped to white and was painted over. No
 * unpainted pixel can wear a mark.
 */
export const CLIP_MARKS: Readonly<Record<Clip, readonly [number, number, number]>> = {
  white: [255, 0, 0],
  black: [0, 128, 255],
};

/** What the pointer is over, said one way for every reader. */
export type Readout =
  | { kind: 'value'; r: number; g: number; b: number }
  | { kind: 'clip'; clip: Clip };

/**
 * An 8-bit pixel read off the stage, as the author should be told it. With
 * the clipping view on, a pixel wearing a mark is read as the clip it marks
 * rather than as the mark's own numbers — `0 128 255` over a crushed shadow
 * would be a fabricated value. A step of tolerance, because a scaled draw
 * can round a mark's inner pixel by one.
 */
export function readoutOf(r: number, g: number, b: number, clipping: boolean): Readout {
  if (clipping) {
    for (const clip of ['white', 'black'] as const) {
      const [mr, mg, mb] = CLIP_MARKS[clip];
      if (Math.abs(r - mr) <= 1 && Math.abs(g - mg) <= 1 && Math.abs(b - mb) <= 1) return { kind: 'clip', clip };
    }
  }
  return { kind: 'value', r, g, b };
}

/** "R 212 · G 180 · B 96", or what the mark means. */
export function readoutLabel(readout: Readout): string {
  if (readout.kind === 'clip') return readout.clip === 'white' ? 'clipped to white' : 'crushed to black';
  const n = (v: number) => String(Math.round(v)).padStart(3, ' ');
  return `R ${n(readout.r)} · G ${n(readout.g)} · B ${n(readout.b)}`;
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Render/Clipping.swift
// Clipping — what has gone to white or to black, pixel by pixel: the colours
// the stage paints it in, and how a readout tells a painted pixel from a
// photographed one. Port of `src/shared/render/clipping.ts`; its RULE
// (`clipWhite`, `clipBlack`, `Clip`, `clipOf`) was already ported beside the
// histogram in `Histogram.swift` and is reused here, never restated.
//
// The rules it keeps:
// - each mark has ONE channel at 255, which is what lets `readoutOf` know it:
//   a photographed pixel within a step of either colour has that channel at
//   254 or more, so it is itself clipped to white and was painted over — no
//   unpainted pixel can wear a mark;
// - with the clipping view on, a pixel wearing a mark is read as the clip it
//   marks, never as the mark's own numbers, which would be a fabricated value.

import Foundation

/// The colours a clipped pixel is painted in on the stage — Lightroom's own
/// (red for highlights, blue for shadows), because a photographer reads them
/// without a legend.
public let clipMarks: [Clip: (r: Int, g: Int, b: Int)] = [
    .white: (255, 0, 0),
    .black: (0, 128, 255),
]

/// What the pointer is over, said one way for every reader.
public enum Readout: Equatable, Sendable {
    case value(r: Double, g: Double, b: Double)
    case clip(Clip)
}

/// An 8-bit pixel read off the stage, as the author should be told it. A step
/// of tolerance, because a scaled draw can round a mark's inner pixel by one.
public func readoutOf(_ r: Double, _ g: Double, _ b: Double, clipping: Bool) -> Readout {
    if clipping {
        for clip in [Clip.white, .black] {
            guard let mark = clipMarks[clip] else { continue }
            if abs(r - Double(mark.r)) <= 1 && abs(g - Double(mark.g)) <= 1 && abs(b - Double(mark.b)) <= 1 {
                return .clip(clip)
            }
        }
    }
    return .value(r: r, g: g, b: b)
}

/// "R 212 · G 180 · B  96", or what the mark means.
public func readoutLabel(_ readout: Readout) -> String {
    switch readout {
    case .clip(let clip):
        return clip == .white ? "clipped to white" : "crushed to black"
    case .value(let r, let g, let b):
        // `Math.round` on a code value, padded to three columns.
        func n(_ v: Double) -> String {
            let s = String(Int(v.rounded()))
            return String(repeating: " ", count: max(0, 3 - s.count)) + s
        }
        return "R \(n(r)) · G \(n(g)) · B \(n(b))"
    }
}
